File: brownfield-backend/src/main/java/com/petclinic/invoice/InvoiceService.java
package com.petclinic.invoice;

import com.lowagie.text.Document;
import com.lowagie.text.Paragraph;
import com.lowagie.text.pdf.PdfWriter;
import org.springframework.stereotype.Service;

import java.io.ByteArrayOutputStream;
import java.util.List;

@Service
public class InvoiceService {

    private final InvoiceRepository invoiceRepository;

    public InvoiceService(InvoiceRepository invoiceRepository) {
        this.invoiceRepository = invoiceRepository;
    }

    public List<Invoice> findAll() {
        return invoiceRepository.findAll();
    }

    public byte[] generatePdf(Long invoiceId) {
        Invoice invoice = invoiceRepository.findById(invoiceId)
                .orElseThrow(() -> new RuntimeException("Invoice not found"));

        try (ByteArrayOutputStream out = new ByteArrayOutputStream()) {
            Document document = new Document();
            PdfWriter.getInstance(document, out);
            document.open();

            document.add(new Paragraph("Invoice Details"));
            document.add(new Paragraph("Invoice Number: " + invoice.getInvoiceNumber()));
            document.add(new Paragraph("Date: " + invoice.getInvoiceDate()));
            document.add(new Paragraph("Amount: $" + invoice.getAmount()));
            
            document.add(new Paragraph("\nVisit Details"));
            if (invoice.getVisit() != null) {
                document.add(new Paragraph("Clinic: " + invoice.getVisit().getClinic()));
                document.add(new Paragraph("Summary: " + invoice.getVisit().getSummary()));
                if (invoice.getVisit().getPet() != null) {
                     document.add(new Paragraph("Pet: " + invoice.getVisit().getPet().getName()));
                }
                if (invoice.getVisit().getVet() != null) {
                     document.add(new Paragraph("Vet: " + invoice.getVisit().getVet().getName()));
                }
            }

            document.close();
            return out.toByteArray();
        } catch (Exception e) {
            throw new RuntimeException("Error generating PDF", e);
        }
    }
}

File: brownfield-backend/src/invoice/invoiceService.ts
import PDFDocument from "pdfkit";
import { Invoice } from "./invoice";
import { InvoiceRepository } from "./invoiceRepository";

export class InvoiceService {
  private invoiceRepository: InvoiceRepository;

  constructor(invoiceRepository?: InvoiceRepository) {
    this.invoiceRepository = invoiceRepository ?? new InvoiceRepository();
  }

  findAll(): Promise<Invoice[]> {
    return this.invoiceRepository.findAll();
  }

  async generatePdf(invoiceId: number): Promise<Buffer> {
    const invoice = await this.invoiceRepository.findById(invoiceId);
    if (!invoice) {
      throw new Error("Invoice not found");
    }

    return new Promise<Buffer>((resolve, reject) => {
      const doc = new PDFDocument();
      const chunks: Buffer[] = [];

      doc.on("data", (chunk: Buffer) => chunks.push(chunk));
      doc.on("end", () => resolve(Buffer.concat(chunks)));
      doc.on("error", reject);

      doc.fontSize(18).text("Invoice Details", { underline: true });
      doc.moveDown();
      doc.fontSize(12).text(`Invoice Number: ${invoice.invoiceNumber}`);
      doc.text(`Date: ${invoice.invoiceDate}`);
      doc.text(`Amount: $${invoice.amount}`);

      if (invoice.visit) {
        doc.moveDown();
        doc.fontSize(14).text("Visit Details", { underline: true });
        doc.fontSize(12).text(`Clinic: ${invoice.visit.clinic}`);
        doc.text(`Summary: ${invoice.visit.summary}`);
        if (invoice.visit.pet) {
          doc.text(`Pet: ${invoice.visit.pet.name}`);
        }
        if (invoice.visit.vet) {
          doc.text(`Vet: ${invoice.visit.vet.name}`);
        }
      }

      doc.end();
    });
  }
}
